### PR TITLE
Add com.ghostery.browser

### PR DESCRIPTION
**Pull request by milouse**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*
  + gtk3-werror.patch is required as part of cups support (also used by librewolf or firefox flatpak. See https://github.com/flathub/org.mozilla.firefox.BaseApp/pull/21 https://github.com/flathub/org.mozilla.firefox.BaseApp/pull/22 and https://phabricator.services.mozilla.com/D138908

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission

Hi there,

We would like to distribute our mozilla-based browser on flathub. This is our first interaction with flathub or flatpak in general, thus do not hesitate to come back to us if we did something wrong. We tried hard to follow all readmes/guidelines, but we might have missed something. This manifest is heavily inspired by the one of librewolf, if it can help review.

Thank you very much!

### CODE REVIEW

File: com.ghostery.browser.yml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  - install -D -m644 /app/lib/ghostery/browser/chrome/icons/default/default128.png
+    /app/usr/share/icons/hicolor/128x128/apps/com.ghostery.browser.png
+  - sed -i 's|"app.update.channel", "release"|"app.update.channel", "flatpak"|' /app/lib/ghostery/defaults/pref/channel-prefs.js
+  - install -D -m644 com.ghostery.browser.desktop -t /app/share/applications
+  - install -D -m644 com.ghostery.browser.metainfo.xml -t /app/share/metainfo
+  - install -D -m755 ghostery.sh /app/bin/com.ghostery.browser
+  - install -d /app/lib/ffmpeg
+  sources:
+  - type: archive
+    url: https://github.com/ghostery/user-agent-desktop/releases/download/2023-07-21/Ghostery-2023.6.en-US.linux.tar.gz
```
Comment (JakobDev) [inline com.ghostery.browser.yml:55]: Why you don't build from source?

File: com.ghostery.browser.yml [hunk 2]
```diff
@@ -0,0 +1,102 @@
+---
+app-id: com.ghostery.browser
+default-branch: stable
```
Comment (hfiguiere) [inline com.ghostery.browser.yml:3]:
```suggestion
```

File: com.ghostery.browser.yml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+- "--talk-name=org.freedesktop.FileManager1"
+- "--system-talk-name=org.freedesktop.NetworkManager"
+- "--talk-name=org.a11y.Bus"
+- "--talk-name=org.gnome.SessionManager"
+- "--talk-name=org.freedesktop.ScreenSaver"
+- "--talk-name=org.gtk.vfs.*"
+- "--talk-name=org.freedesktop.Notifications"
+- "--env=MOZ_ENABLE_WAYLAND=1"
+- "--env=MOZ_USE_XINPUT2=1"
+- "--own-name=com.ghostery.browser.*"
```
Comment (hfiguiere) [inline com.ghostery.browser.yml:35]:
```suggestion
```

this is granted based on the app-id

File: com.ghostery.browser.yml [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+- "--own-name=org.mpris.MediaPlayer2.firefox.*"
```
Comment (hfiguiere) [inline com.ghostery.browser.yml:36]: you mean that it doesn't use it's own mpris name and "borrow" the one from Firefox?

File: com.ghostery.browser.yml [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+modules:
+- shared-modules/dbus-glib/dbus-glib-0.110.json
+- name: ghostery_browser
```
Comment (hfiguiere) [inline com.ghostery.browser.yml:39]: it's customary to have the main module last because it's dependency order.

File: com.ghostery.browser.yml (hunk 1 above)
Comment (hfiguiere) [inline com.ghostery.browser.yml:55]: how does this work with aarch64 since an aarch64 package is build here as well?

**Comment by hfiguiere** [comment]
bot, build com.ghostery.browser

File: com.ghostery.browser.yml [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  buildsystem: simple
+  build-commands:
+  - mkdir -p /app/lib
+  - mv ghostery_app/Ghostery /app/lib/ghostery
+  - install -D -m644 /app/lib/ghostery/browser/chrome/icons/default/default128.png
+    /app/share/icons/hicolor/128x128/apps/com.ghostery.browser.png
+  - install -D -m644 /app/lib/ghostery/browser/chrome/icons/default/default128.png
+    /app/usr/share/icons/hicolor/128x128/apps/com.ghostery.browser.png
```
Comment (hfiguiere) [inline com.ghostery.browser.yml:47]: why do you install an icon in `/app/usr` ?

File: com.ghostery.browser.yml (hunk 3 above)
Comment (hfiguiere) [reply com.ghostery.browser.yml:35]: this is where the linter errors.

Comment (milouse) [reply com.ghostery.browser.yml:35]: Oh, I didn’t know. Thanks for the tip!

File: com.ghostery.browser.yml (hunk 1 above)
Comment (milouse) [reply com.ghostery.browser.yml:55]: Can you explain why aarch64 will be built too? I added the `only-arches: [x86_64]` attribute to this source to only build for this target. Did I misunderstood something?

Comment (hfiguiere) [reply com.ghostery.browser.yml:55]: this just exclude the *module*

https://docs.flathub.org/docs/for-app-authors/maintanance/#limiting-the-set-of-architectures-to-build-on

But then why not building from source to ship an aarch64 package.

File: com.ghostery.browser.yml (hunk 4 above)
Comment (milouse) [reply com.ghostery.browser.yml:36]: Yes, [same as for LibreWolf](https://github.com/flathub/io.gitlab.librewolf-community/blob/master/io.gitlab.librewolf-community.json#L37). It’s a limitation in the mozilla build chain. As we don’t want to patch it too deeper, we keep this naming.

File: com.ghostery.browser.yml (hunk 6 above)
Comment (milouse) [reply com.ghostery.browser.yml:47]: It looks like another misunderstanding. I removed it.

File: com.ghostery.browser.yml (hunk 1 above)
Comment (milouse) [reply com.ghostery.browser.yml]: Please see my general comment below

Comment (milouse) [reply com.ghostery.browser.yml]: Got it.

About the build step, please see my general comment below.

**Comment by milouse** [comment]
Thank you very much for your quick and detailed review. I hope my various answer/action will be ok for you.

Both of you ask about why not actually build the browser when packaging it. The mozilla build chain is a bit complicated and requires both big resources and lot of various customization. From our point of view it’s way more simpler to only build our browser once on our side, controlling the whole process, then distributing it on various channel from the same tarball. It looks a bit as a waste of resources to compile it several time, just for various packaging purpose. This also simplify a lot our support process, knowing that normally our end-users will have the same binary under the hood. So for now, if it’s not a problem we would like to keep things like this. However we do not completely close the door to build a specific version of our browser while creating flatpak package. But currently we do not have the resources to do it.

**Comment by JakobDev** [comment]
Please add a flathub.json file with the following content:
```json
{
    "only-arches": ["x86_64"]
} 
```
This  disables the ARM build.

A problem I can see is that you have different binaries for different languages (en,de,fr). So if a User installs Ghostery from Flathub, he will only get the english version. Could you pack all languages into one binary?

**Comment by milouse** [comment]
Thank you for your feedback!

> Please add a flathub.json file with the following content:
> 
> ```json
> {
>     "only-arches": ["x86_64"]
> } 
> ```
> 
> This disables the ARM build.

Done in my last commit.

> A problem I can see is that you have different binaries for different languages (en,de,fr). So if a User installs Ghostery from Flathub, he will only get the english version. Could you pack all languages into one binary?

Yes, we know about it and it’s on our roadmap. For various reason we did not succeed to do it for now (nor took time to fix it). So yes our first flatpak release(s) will be english only. It will be fixed on a future release.

**Comment by hfiguiere** [comment]
bot, build com.ghostery.browser

File: com.ghostery.browser.yml [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  - install -D -m755 ghostery.sh /app/bin/com.ghostery.browser
+  - install -d /app/lib/ffmpeg
+  sources:
+  - type: archive
+    url: https://github.com/ghostery/user-agent-desktop/releases/download/2023-07-21/Ghostery-2023.6.en-US.linux.tar.gz
+    sha256: 84bbd2221172e4b78fc88c8f9c46af0a43c8cc9e08d33ec32e7e35334b9e8609
+    dest: ghostery_app
+    strip-components: 0
+  - type: file
+    path: "../data/ghostery-private-browser.desktop"
```
Comment (hfiguiere) [inline com.ghostery.browser.yml:92]:
```
flatpak-builder: Downloading ghostery_browserFailed to download sources: module ghostery_browser: Unable to get source file '../data/ghostery-private-browser.desktop': File '/srv/buildbot/worker/build-x86_64-2/data' not found
```

File: com.ghostery.browser.yml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    dest-filename: com.ghostery.browser.desktop
```
Comment (hfiguiere) [inline com.ghostery.browser.yml:93]:
```suggestion
    path: com.ghostery.browser.desktop
```

Comment (milouse) [reply com.ghostery.browser.yml:93]: 🤦 exactly. Very sorry we are using another repo to test flatpak, in which the desktop file is shared in another folder. When I moved the ghostery target from the second to the last step, I forget to fix the file path :/ My bad. I’ll fix that.

**Comment by milouse** [comment]
Done. Sorry for the back and forth :/

**Comment by hfiguiere** [comment]
bot, build com.ghostery.browser
